dagutils/traversal.go

```diff
@@ -0,0 +1,507 @@
+	// would require a special visit case inside every traversal operation like
+	// `Iterate()` and `Search`). This value should never be returned to after
+	// the first `down` movement, moving up from the root should always return
+	// `ErrUpOnRoot`.
+	level int
+
+	// Method called each time a node is arrived upon in a traversal (through
+	// the `down()` movement, that is, when visiting it for the first time,
+	// not when going back up). It is the main API to implement functionality
+	// (like read and seek) on top of the `Traversal` structure.
+	//
+	// Its argument is the current `node` being visited. Any error it returns
+	// (apart from the internal `errDownStopIteration`) will be forwarded to
+	// the caller of the traversal operation (stopping it).
+	//
+	// Any of the exported methods of this API should be allowed to be called
+	// from within this method.
+	VisitHandler func(node ipld.Node) error
+
+	// Flag to stop the current iteration, intended to be set by the user
+	// inside `VisitHandler` to stop the current traversal operation.
+	Stop bool
+	// TODO: Use a function?
+
+	// Attribute needed to fetch nodes, will become really useful once node
+	// promises are implemented.
+	ctx  context.Context
+	serv ipld.NodeGetter
+
+	// The CID of each child of each node in the `path` (indexed by
+	// `level` and `childIndex`).
+	childCIDs [][]*cid.Cid
+
+	// NodePromises for child nodes requested for every node in the
+	// `path` (as `childCIDs`, indexed by `level` and `childIndex`).
+	promises [][]*ipld.NodePromise
+	// TODO: Consider encapsulating in a single structure along `childCIDs`.
+
+	// Cancel methods for every node in the `path` that had requested
+	// child nodes through the `promises`.
+	cancel []func()
+}
+
+// NewTraversal creates a new `Traversal` structure from a `root`
+// IPLD node.
+func NewTraversal(ctx context.Context, root ipld.Node, serv ipld.NodeGetter) *Traversal {
+	return &Traversal{
+		path:       []ipld.Node{root},
+		childIndex: []uint{0},
+		level:      -1,
+		// Starting position, "on top" of the root node, see `level`.
+
+		ctx:  ctx,
+		serv: serv,
+
+		childCIDs: make([][]*cid.Cid, 1),
+		promises:  make([][]*ipld.NodePromise, 1),
+		cancel:    make([]func(), 1),
+		// Initial capacity of 1 (needed for the doubling capacity algorithm
+		// of `extendPath`, it can't be zero).
+	}
+}
+
+// ErrDownNoChild signals there are no more child nodes left to visit,
+// the current child index is past the end of this node's links.
+var ErrDownNoChild = errors.New("can't go down, no child available")
+
+// ErrUpOnRoot signals the end of the DAG after returning to the root.
+var ErrUpOnRoot = errors.New("can't go up, already on root")
+
+// ErrRightNoChild signals the end of this parent child nodes.
+var ErrRightNoChild = errors.New("can't move right, no more child nodes in this parent")
+
+// errDownStopIteration signals the stop of the traversal operation.
+var errDownStopIteration = errors.New("stop")
```

Comment:
I wonder if `Pause` would be better than `Stop` here — in my mind, stopping means you can’t continue from where you left off. With this code, though, you can (and that’s important, because that’s how seeking works).